### PR TITLE
[repo] Update System.Text.Json to 8.0.4

### PR DESCRIPTION
**Pull request by joegoldman2**
## Changes

Update System.Text.Json to 8.0.4 to mitigate CVE-2024-30105.
Addresses https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2125#discussion_r1785214763.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.Exporter.OneCollector/OpenTelemetry.Exporter.OneCollector.csproj [hunk 1]
```diff
@@ -17,12 +17,11 @@
   </PropertyGroup>
 
   <PropertyGroup>
-    <OTelSdkVersion>$(OpenTelemetryCoreLatestVersion)</OTelSdkVersion>
-    <DefineConstants Condition="$(OTelSdkVersion.Contains('alpha')) OR $(OTelSdkVersion.Contains('beta')) OR $(OTelSdkVersion.Contains('rc'))">$(DefineConstants);EXPOSE_EXPERIMENTAL_FEATURES</DefineConstants>
```
Comment (Kielek) [inline src/OpenTelemetry.Exporter.OneCollector/OpenTelemetry.Exporter.OneCollector.csproj:21]: Separate PR. There is more occurrences of this env.

File: src/OpenTelemetry.Resources.Gcp/OpenTelemetry.Resources.Gcp.csproj [hunk 2]
```diff
@@ -15,9 +15,9 @@
   <ItemGroup>
     <PackageReference Include="Google.Api.Gax" Version="4.8.0" />
     <PackageReference Include="OpenTelemetry" Version="$(OpenTelemetryCoreLatestVersion)" />
-    <PackageReference Include="System.Text.Json" Version="4.7.2" />
+    <PackageReference Include="System.Text.Json" Version="$(SystemTextJsonPkgVer)" />
     <!-- System.Text.Encodings.Web is indirect reference. It is needed to upgrade it directly to avoid https://github.com/advisories/GHSA-ghhp-997w-qr28 -->
-    <PackageReference Include="System.Text.Encodings.Web" Version="4.7.2" />
+    <PackageReference Include="System.Text.Encodings.Web" Version="$(SystemTextEncodingsWebPkgVer)" />
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Resources.Gcp/OpenTelemetry.Resources.Gcp.csproj:20]: You can remove the `System.Text.Encodings.Web` reference where you have a direct reference to `System.Text.Json` v8.0.4. The reason `System.Text.Encodings.Web` was here is the System.Text.Json v4.7.2 requires `System.Text.Encodings.Web` >= 4.7.1 but `System.Text.Encodings.Web` v4.7.1 is vulnerable so you have to explicitly add a reference to v4.7.2.

When you reference `System.Text.Json` v8.0.4 you get `System.Text.Encodings.Web` v8.0.0 transitively and it is safe (for the moment) 😏

File: src/OpenTelemetry.Extensions.AWS/CHANGELOG.md [hunk 3]
```diff
@@ -15,6 +15,10 @@ Released 2024-Sep-24
 * Drop support for .NET 6 as this target is no longer supported and add .NET 8 target.
   ([#2125](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2125))
 
+* Update `System.Text.Json` to `8.0.4` to mitigate
+  [CVE-2024-30105](https://github.com/advisories/GHSA-hh2w-p6rv-4g7w).
+  ([#2149](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2149))
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions.AWS/CHANGELOG.md:20]: This is a lie for this project, not sure about others.

The reason is the `Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'"` in here...

```xml
<PackageReference Include="System.Text.Json" Version="$(SystemTextJsonPkgVer)" Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'" />
```

`net8.0` is going to stay on the framework reference version which may be 8.0.0 - 8.0.3.

What we have to do is add a direct reference to System.Text.Json when the TFM is `net8.0`.

That's why on the main repo PR I did the targets:

```xml
  <ItemGroup Condition="'$(ReferenceSystemTextJsonPackages)' == 'true' AND $(TargetFrameworksRequiringSystemTextJsonDirectReference.Contains('$(TargetFramework)'))">
    <PackageReference Include="System.Text.Json" />
  </ItemGroup>
````

File: src/OpenTelemetry.Resources.Gcp/OpenTelemetry.Resources.Gcp.csproj (hunk 2 above)
Comment (joegoldman2) [reply src/OpenTelemetry.Resources.Gcp/OpenTelemetry.Resources.Gcp.csproj]: I didn't notice `System.Text.Json` depends on `System.Text.Encodings.Web`. I removed direct reference to `System.Text.Encodings.Web` in this case.

File: src/OpenTelemetry.Extensions.AWS/CHANGELOG.md (hunk 3 above)
Comment (joegoldman2) [reply src/OpenTelemetry.Extensions.AWS/CHANGELOG.md:20]: I added the reference to `System.Text.Json` for all targets in this package as it is already the case for all others like:

https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/e5106b25f96815ec0c75e912bdf071f7877bd8f0/src/OpenTelemetry.Resources.Gcp/OpenTelemetry.Resources.Gcp.csproj#L18

Let me know if it is ok for you.

**Review by reyang — CHANGES_REQUESTED** [review]
This is a breaking change, need to provide impact analysis.
For example, customers who are using https://www.nuget.org/packages/System.Text.Json version v6.0.x will forced to upgrade to v8.0.4 (which has no backward compatibility guarantee bewteen v8 and v6), although v6.0.x doesn't have such security vulnerability.

**Comment by joegoldman2** [comment]
> This is a breaking change, need to provide impact analysis. For example, customers who are using https://www.nuget.org/packages/System.Text.Json version v6.0.x will forced to upgrade to v8.0.4 (which has no backward compatibility guarantee bewteen v8 and v6), although v6.0.x doesn't have such security vulnerability.

Ok I understand. Will have to think how to address https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2125#discussion_r1785214763 in this case.

**Review by ppittle — APPROVED** [review]

File: src/OpenTelemetry.Exporter.OneCollector/CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,10 @@
 * Drop support for .NET 6 as this target is no longer supported.
   ([#2123](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2123))
 
+* Update `System.Text.Json` to `8.0.4` to mitigate
+  [CVE-2024-30105](https://github.com/advisories/GHSA-hh2w-p6rv-4g7w).
```
[the remark this answers is not in the document]
Comment (joegoldman2) [reply src/OpenTelemetry.Exporter.OneCollector/CHANGELOG.md:9]: The changelog, at this point, is correct as this package is using a global variable that has been updated.

**Comment by joegoldman2** [comment]
Could the following logic be okay?
- If `$(TargetFrameworkIdentifier)` is `.NETFramework` or `$(TargetFrameworkIdentifier)` is `.NETStandard` or `$(TargetFramework)` is `net6.0` => `6.0.10` (all versions 6.0.x below are vulnerable)
- If `$(TargetFramework)` is `net8.0` => `8.0.5` (all versions 8.0.x below are vulnerable)

**Comment by Kielek** [comment]
@joegoldman2, thanks for contribution. @CodeBlanch created set of PRs to mitigate issues also in System.Text.Json 8.0.4 (advisory was published yesterday). Leftover should be handled shortly.

**Pull request closed without merging**